### PR TITLE
DRIVERS-555 Unified test format changes for CSOT

### PR DESCRIPTION
**Pull request by divjotarora**
This PR is a rebased and updated version of #873. It only contains Unified Test Format changes and modifications/additions to tests in the `valid-pass` and `invalid` directories. The actual CSOT spec tests will be introduced in another PR to avoid making this one larger than necessary.

Notes for reviewers:

1. This PR bumps the schema version in the unified test format to 1.8 because 1.9
2. Many of the existing tests for `collectionData` in the `invalid` directory weren't functioning as intended because they had an additional `foo` property in the collection entity declaration. As a result, they were considered invalid because the collection entity was not valid and the `initialData` field was not being checked. These have been fixed in this PR.
3. ~The aforementioned `collectionData` test files were renamed from `collectionData*` to `initialCollectionData*` to reflect a similar change in the test format spec. A new set of `outcomeCollectionData*` tests have also been added.~

### CODE REVIEW

File: source/unified-test-format/tests/invalid/expectedError-isTimeoutError-type.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/expectedError-isTimeoutError-type.yml:3]: I noticed you used `schemaVersion: "1.5"` in `initialCollectionData-additionalProperties.yml` below. Did you also intend to use that here since `isTimeoutError` is new syntax?

I realize it makes no difference, since this file is validated against the most recent schema anyway -- but if that ever changes it'd be helpful to have them be in sync.

File: source/unified-test-format/tests/invalid/initialCollectionData-additionalProperties.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/initialCollectionData-additionalProperties.yml:3]: I don't think anything here pertains to newer schema versions, since renaming `collectionData` to `initialCollectionData` is strictly internal refactoring. I propose leaving this as 1.0.

File: source/unified-test-format/tests/invalid/initialCollectionData-collectionName-required.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/initialCollectionData-collectionName-required.yml:3]: Likewise, this can remain "1.0".

File: source/unified-test-format/tests/invalid/initialCollectionData-collectionName-type.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/initialCollectionData-collectionName-type.yml:3]: "1.0".

File: source/unified-test-format/schema-1.5.json
Comment (jmikola) [inline source/unified-test-format/schema-1.5.json:200]: Shouldn't this reference `collectionOrDatabaseOptions`?

File: source/unified-test-format/tests/invalid/initialCollectionData-databaseName-required.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/initialCollectionData-databaseName-required.yml:3]: "1.0"

File: source/unified-test-format/tests/invalid/initialCollectionData-databaseName-type.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/initialCollectionData-databaseName-type.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/initialCollectionData-documents-items.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/initialCollectionData-documents-items.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/initialCollectionData-documents-required.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/initialCollectionData-documents-required.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/initialCollectionData-documents-type.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/initialCollectionData-documents-type.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/outcomeCollectionData-additionalProperties.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/outcomeCollectionData-additionalProperties.yml:3]: This can also be "1.0", as it would have been invalid based on the original schema when `collectionData` was a shared definition.

File: source/unified-test-format/tests/invalid/outcomeCollectionData-collectionName-required.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/outcomeCollectionData-collectionName-required.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/outcomeCollectionData-collectionName-type.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/outcomeCollectionData-collectionName-type.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/outcomeCollectionData-databaseName-required.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/outcomeCollectionData-databaseName-required.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/outcomeCollectionData-databaseName-type.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/outcomeCollectionData-databaseName-type.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/outcomeCollectionData-documents-items.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/outcomeCollectionData-documents-items.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/outcomeCollectionData-documents-required.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/outcomeCollectionData-documents-required.yml:3]: "1.0".

File: source/unified-test-format/tests/invalid/outcomeCollectionData-documents-type.yml
Comment (jmikola) [inline source/unified-test-format/tests/invalid/outcomeCollectionData-documents-type.yml:3]: "1.0".

File: source/unified-test-format/tests/valid-pass/entity-cursor-iterateOnce.yml
Comment (jmikola) [inline source/unified-test-format/tests/valid-pass/entity-cursor-iterateOnce.yml:41]: Thank you for the comment.

File: source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml
Comment (jmikola) [inline source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml:22]: Would it make sense for this test to do something more to assert that `collectionOptions` were actually applied when creating the collection? As-is, I'd worry this could pass even if a driver ignored `collectionOptions`.

I realize we can't observe APM events for this, but perhaps you could use [`$collStats`](https://docs.mongodb.com/manual/reference/operator/aggregation/collStats/#storagestats-document) aggregation stage (with a 3.4+ server version requirement) or the [`collStats`](https://docs.mongodb.com/manual/reference/command/collStats/) command and match the result with `capped` and `size`.

File: source/unified-test-format/tests/valid-pass/matches-lte-operator.yml
Comment (jmikola) [inline source/unified-test-format/tests/valid-pass/matches-lte-operator.yml:36]: To test the boundary case for `$$lte`, I'd suggest inserting `{x: 1, y: 1}` and then matching against `{ x: { $$lte: 1 }, y: { $$lte: 2 }}`.

File: source/unified-test-format/unified-test-format.rst
Comment (jmikola) [inline source/unified-test-format/unified-test-format.rst:1780]: For the benefit of others that might review this, I'll note that SHOULD is being used because the schema does not enforce this so the best we can do is ask folks not to write such tests.

File: source/unified-test-format/unified-test-format.rst
Comment (jmikola) [inline source/unified-test-format/unified-test-format.rst:736]: Can we link to https://docs.mongodb.com/manual/reference/command/create/ here? Since we don't have a collection management spec (as we do with indexes), that's going to be the primary resource for anyone trying to determine what options are supported.

I would also add a line that tests SHOULD NOT utilize the `writeConcern` option, since that could conflict with the "majority" write concern we require later on (feel free to note that as the justification).

File: source/unified-test-format/unified-test-format.rst
Comment (jmikola) [inline source/unified-test-format/unified-test-format.rst:349]: I've noted that this is redundant given what is already specified in "Executing a Test".

File: source/unified-test-format/unified-test-format.rst
Comment (jmikola) [inline source/unified-test-format/unified-test-format.rst:2856]: This predates your PR, but since you're changing this please insert "and" for `, and a "local" read concern.`.

**Review by jmikola — COMMENTED** [review]
Don't forget version bump and changlog entries in the spec RST.

File: source/unified-test-format/unified-test-format.rst
Comment (divjotarora) [reply source/unified-test-format/unified-test-format.rst]: Added the link and a note about `writeConcern`.

File: source/unified-test-format/unified-test-format.rst
Comment (divjotarora) [reply source/unified-test-format/unified-test-format.rst]: Done.

File: source/unified-test-format/schema-1.5.json
Comment (jmikola) [reply source/unified-test-format/schema-1.5.json:200]: Thanks. I meant to delete this comment before submitting the review, as I answered the question for myself when I read further along in the PR (see my other comment about advising against using `writeConcern`).

File: source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml
Comment (jmikola) [inline source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml:1]: I think this should just be `initialCollectionData-collectionOptions`. Please double-check that none of the other files have the extension in their description.

File: source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml
Comment (jmikola) [inline source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml:33]: I know #944 was closed and moved internally for the time being, but heads up to @patrickfreed in case this test will need to be skipped for serverless. IIRC, `$collStats` isn't supported but I'm not sure about the generic command.

Perhaps there's a more general concern with tests running arbitrary commands that might not be in the versioned API, which will required auditing during the course of that project.

File: source/unified-test-format/tests/valid-pass/matches-lte-operator.yml
Comment (jmikola) [inline source/unified-test-format/tests/valid-pass/matches-lte-operator.yml:37]: I'm not sure if this was a leftover comment, but if not: what is "exact assertions" referring to? Are you implying that this is incomplete because we cannot assert a case where `$$lte`` fails to match? Presumably we could do that with a `valid-fail` test, although we wouldn't be able to confirm the exact cause of failure.

File: source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml
Comment (divjotarora) [reply source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml:1]: Done.

File: source/unified-test-format/tests/valid-pass/matches-lte-operator.yml
Comment (divjotarora) [reply source/unified-test-format/tests/valid-pass/matches-lte-operator.yml:37]: The comment is intended to say that we could perform an exact comparison with the document `{_id: 1, y: 1}` but we use `$$lte` instead to test support.

File: source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml
Comment (divjotarora) [reply source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml:33]: Thanks for the heads up. I made a note to investigate this and skip any of the tests added in this PR or in #960 if they're incompatible with serverless.

File: source/unified-test-format/tests/valid-pass/matches-lte-operator.yml
Comment (jmikola) [reply source/unified-test-format/tests/valid-pass/matches-lte-operator.yml:37]: Oh, that makes more sense.

**Review by jmikola — APPROVED** [review]

File: source/unified-test-format/unified-test-format.rst
Comment (DmitryLukyanov) [inline source/unified-test-format/unified-test-format.rst:1782]: >If the cursor's current batch is empty, one ``getMore`` MUST be attempted to get more results

so, this means that we will iterate until a document like `iterateUntilDocumentOrError`. Then the only difference that this step is not expected to fail?

**Review by DmitryLukyanov — APPROVED** [review]
One question and then LGTM

File: source/unified-test-format/unified-test-format.rst
Comment (divjotarora) [reply source/unified-test-format/unified-test-format.rst:1782]: The difference is that even if the `getMore` returns an empty batch, the driver ends the operation with an empty result. This is useful for cases like change streams. IIRC, it's used in a change streams test to assert that a `getMore` was sent and the timeout was applied to the command but we don't care about the result.

Comment (DmitryLukyanov) [reply source/unified-test-format/unified-test-format.rst:1782]: then we have a bug in `IterateUntilDocumentOrErrorOperation`: https://github.com/mongodb/mongo-csharp-driver/blob/master/tests/MongoDB.Driver.Tests/UnifiedTestOperations/UnifiedIterateUntilDocumentOrErrorOperation.cs#L38, the result can contain an empty batch, so we don't wait until the document appearance

**Comment by benjirewis** [comment]
I wonder if we should add a `valid-pass` test testing the functionality of the new `createEntities` operation as distinct from the file-level `createEntities`.

File: source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml
Comment (benjirewis) [inline source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml:39]: `collection is created with the correct options` fails in the Go driver on server versions 2.6 and 3.0. In 2.6, `maxSize` [is not returned](https://evergreen.mongodb.com/lobster/evergreen/test/mongo_go_driver_tests_legacy_auth_ssl__version~2.6_os_ssl_legacy~ubuntu1404_go_1_16_test_standalone_noauth_nossl_patch_0b0acdc5f38771ecf4d768179a315fe919d9f4b0_6259cfc42fbabe7dcb08d893_22_04_15_20_04_22/0/test/#bookmarks=0%2C39&l=1&shareLine=18) in the `collStats` command, and in `3.0`, `maxSize` [is 4096](https://evergreen.mongodb.com/lobster/evergreen/test/mongo_go_driver_tests_legacy_auth_ssl__version~3.0_os_ssl_legacy~ubuntu1404_go_1_16_test_standalone_noauth_nossl_patch_0b0acdc5f38771ecf4d768179a315fe919d9f4b0_6259cfc42fbabe7dcb08d893_22_04_15_20_04_22/0/test/#bookmarks=0%2C39&l=1&shareLine=18) (the default) instead of the expected 512. We could change this test to have a `minServerVersion: "3.6"` unless others are against.

Comment (benjirewis) [inline source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml:39]: `collection is created with the correct options` fails in the Go driver on server versions 2.6 and 3.0. In 2.6, `maxSize` [is not returned](https://evergreen.mongodb.com/lobster/evergreen/test/mongo_go_driver_tests_legacy_auth_ssl__version~2.6_os_ssl_legacy~ubuntu1404_go_1_16_test_standalone_noauth_nossl_patch_0b0acdc5f38771ecf4d768179a315fe919d9f4b0_6259cfc42fbabe7dcb08d893_22_04_15_20_04_22/0/test/#bookmarks=0%2C39&l=1&shareLine=18) in the `collStats` command, and in `3.0`, `maxSize` [is 4096](https://evergreen.mongodb.com/lobster/evergreen/test/mongo_go_driver_tests_legacy_auth_ssl__version~3.0_os_ssl_legacy~ubuntu1404_go_1_16_test_standalone_noauth_nossl_patch_0b0acdc5f38771ecf4d768179a315fe919d9f4b0_6259cfc42fbabe7dcb08d893_22_04_15_20_04_22/0/test/#bookmarks=0%2C39&l=1&shareLine=18) (the default) instead of the expected 512. We could change this test to have a `minServerVersion: "3.6"` unless others are against.

**Comment by ShaneHarvey** [comment]
> 
I wonder if we should add a valid-pass test testing the functionality of the new createEntities operation as distinct from the file-level createEntities.

Done. I also resolved the merge conflicts and updated the spec version to 1.8. The unified tests are passing in Python.

File: source/unified-test-format/tests/valid-pass/createEntities-operation.yml
Comment (benjirewis) [inline source/unified-test-format/tests/valid-pass/createEntities-operation.yml:3]: What is in schema version `1.8`? The `createEntities` operation?

**Review by benjirewis — APPROVED** [review]
Question about schema version but otherwise LGTM; new test works in the Go driver and seems to test the correct functionality.

File: source/unified-test-format/tests/valid-pass/entity-cursor-iterateOnce.yml
Comment (jmikola) [inline source/unified-test-format/tests/valid-pass/entity-cursor-iterateOnce.yml:3]: Similar to my earlier suggestion:

```suggestion
# Note: iterateOnce is not technically in the 1.8 schema is implemented alongside related changes
schemaVersion: "1.8"
```

File: source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml
Comment (jmikola) [reply source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml:36]: Bumping this previously resolved comment.

[Unsupported Commands in Serverless Instances](https://www.mongodb.com/docs/atlas/unsupported-commands/#std-label-atlas-serverless-command-limitations) doesn't include `collStats` so I think we're fine here. Just a reminder to run these tests through Atlas Serverless if you get a chance so we double-check before merging. Feel free to resolve after doing so.

File: source/unified-test-format/tests/valid-pass/matches-lte-operator.yml
Comment (jmikola) [inline source/unified-test-format/tests/valid-pass/matches-lte-operator.yml:3]: Feel free to ignore if you think these comments are tedious.

```suggestion
# Note: $$lte is not technically in the 1.8 schema is implemented alongside related changes
schemaVersion: "1.8"
```

File: source/unified-test-format/unified-test-format.rst
Comment (jmikola) [inline source/unified-test-format/unified-test-format.rst:6]: Note: https://github.com/mongodb/specifications/pull/1188 is also bumping the schema version so depending on which PR gets merged first the other will need to rebase accordingly.

File: source/unified-test-format/schema-1.8.json
Comment (jmikola) [inline source/unified-test-format/schema-1.8.json:205]: On a separate note, I think you can actually enforce the absence of a `writeConcern` option within `collectionOptions`. See: https://stackoverflow.com/a/30725034

File: source/unified-test-format/unified-test-format.rst
Comment (jmikola) [inline source/unified-test-format/unified-test-format.rst:767]: Per my earlier comment in the JSON schema, I think you can enforce this (see: https://stackoverflow.com/a/30725034) and use `MUST NOT` here if that's preferable.

File: source/unified-test-format/unified-test-format.rst
Comment (jmikola) [inline source/unified-test-format/unified-test-format.rst:2244]: For consistency with what we actually do in most existing tests.

```suggestion
              databaseName: &database0Name test
```

**Review by ShaneHarvey — COMMENTED** [review]
Rebased and updated to schema version 1.9 with the suggested changes.

File: source/unified-test-format/unified-test-format.rst
Comment (ShaneHarvey) [reply source/unified-test-format/unified-test-format.rst]: Done.

File: source/unified-test-format/schema-1.8.json
Comment (ShaneHarvey) [reply source/unified-test-format/schema-1.8.json]: Done.

File: source/unified-test-format/tests/valid-pass/createEntities-operation.yml
Comment (jmikola) [reply source/unified-test-format/tests/valid-pass/createEntities-operation.yml:31]: Only benefit is if other spec authors use these as examples down the line. I have no objection to leaving this as-is.

Comment (ShaneHarvey) [reply source/unified-test-format/tests/valid-pass/createEntities-operation.yml:31]: To be clear, I don't understand what benefit there is to using anchors for entity IDs in any test at all (not just this one).

File: source/unified-test-format/schema-1.9.json
Comment (ShaneHarvey) [reply source/unified-test-format/schema-1.9.json:284]: I merged initialCollectionData and outcomeCollectionData based on your suggestion in https://github.com/mongodb/specifications/pull/959#discussion_r858944692

Did I misunderstand? Do you think I should revert back to separate entities?

Comment (jmikola) [reply source/unified-test-format/schema-1.9.json:284]: Those were separate points and my only suggestion was the `createOptions` rename; however, I don't object to keeping them consolidated as you have now.

File: source/unified-test-format/tests/valid-pass/createEntities-operation.yml
Comment (jmikola) [reply source/unified-test-format/tests/valid-pass/createEntities-operation.yml:31]: In the case of entity IDs where the anchor and value are almost always identical, I can see why it seems pointless. I think one small benefit is that a typo is more likely to be caught during JSON conversion instead of a runtime error when executing the tests (due to a failed entity lookup).

For database and collection names, I think the anchors serve a much greater purpose since the values are often very different from the anchor names.

File: source/unified-test-format/schema-1.9.json
Comment (ShaneHarvey) [reply source/unified-test-format/schema-1.9.json:285]: Okay thanks, I've kept them combined and fixed the duplicate test file issue.

**Comment by ShaneHarvey** [comment]
I resolved the merge conflicts and cleaned up the duplicate initialCollectionData and outcomeCollectionData tests. Ready for another look.

File: source/unified-test-format/tests/valid-pass/createEntities-operation.yml
Comment (ShaneHarvey) [reply source/unified-test-format/tests/valid-pass/createEntities-operation.yml:31]: Gotcha. Catching typos does seem like a useful benefit.

File: source/unified-test-format/tests/valid-pass/createEntities-operation.yml
Comment (jmikola) [inline source/unified-test-format/tests/valid-pass/createEntities-operation.yml:18]: Conceivably, you could delete the top-level `createEntities` and `initialData` blocks in this test. They don't appear to be used for anything.

**Review by jmikola — APPROVED** [review]
LGTM. Optional suggestion to remove some lines from `createEntities-operation.yml`.

File: source/unified-test-format/tests/valid-pass/createEntities-operation.yml
Comment (ShaneHarvey) [reply source/unified-test-format/tests/valid-pass/createEntities-operation.yml]: Done.

File: source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml
Comment (jmikola) [reply source/unified-test-format/tests/valid-pass/initialCollectionData-collectionOptions.yml]: Not required by [serverless testing](https://github.com/mongodb/specifications/tree/master/source/serverless-testing#existing-spec-tests), so totally up to you.

I don't think the page I linked above existed when this PR was originally opened, so I just wanted to circle back and cloe the loop on this thread in case anyone comes across it down the line.

**Pull request merged**
